### DISCUSSION TITLE
Compound primary key support - effort to implement?

### DISCUSSION THREAD

**Discussion by sykesd** [malloydata/malloy#788] [Developing Malloy]
What would the effort be to implement support for compound (multi-column) primary keys?

We have a large line-of-business application developed over many years that makes pervasive use of compound primary keys. The ideas behind Malloy are something I whole-heartedly agree with and have pondered implementing (way less elegantly!) myself many times, to help with querying the data in our system.

As far as I can tell from the documentation and inspecting the [grammar](https://github.com/looker-open-source/malloy/blob/main/packages/malloy/src/lang/grammar/Malloy.g4), this is not currently supported. Would there be any interest in adding support, or helping me implement support myself?

As far as I can tell, the `join with` syntax should already support a multi-field foreign key expression using the `ungroup` branch of `fieldExpr`. Whether or not the semantics are implemented is another question.

PS *We know that multi-column primary keys are frowned upon, but this is a large, existing system running a multi-national business where the effort to change this is just not worth it. Yes, if we started again, this is one of those decisions that would be reviewed, but we are not starting again.*

**Comment by anikaks** [comment]
Hey @sykesd , a couple things here! if you create a dimension that concatenates the parts of the key, you can set that to be the primary key:

```
source: aircraft is table('malloy-data.faa.aircraft') {
  dimension: my_compound_primary_key is concat(tail_num,year_built)
  primary_key: my_compound_primary_key
}
```

As for joining, you can then either reference the above field, and use `join ... with`. This is a silly join to make, but hopefully illustrates how this works:

```
source: flights is table('malloy-data.faa.aircraft') {
  dimension: my_compound_aircraft_key is concat(tail_num,dep_time.year)
  join_one: aircrafts with my_compound_aircraft_key
}
```


or you can also use `join ... on`:
```
source: flights is table('malloy-data.faa.aircraft') {
  join_one: aircraft on tail_num = aircraft.tail_num and aircraft.year_built = dep_time.year
}
```

Let me know if I'm missing something you were hoping to be able to support here, but I _think_ this should cover you!

**Reply by sykesd** [reply]
Thanks for that. I will certainly give it a try.

Does the `primary key` clause in the `source` do anything other than make the more convenient `join` syntax possible?

**Comment by lloydtabb** [comment]
It gets used in symmetric aggregate calculations which is more efficient
than generating a distinct key.  Symmetric aggregate calculations occur
when computing something that is not in the main table for when the main
table fans out.

**Reply by sykesd** [reply]
Thanks for that. It is going to take me some time to unpack what you said. I have just started reading up on symmetric aggregate calculations. Once I have a better grip on that I can see whether or not it is a concrete issue in our schema.
